### PR TITLE
Add additional client helpers

### PR DESCRIPTION
**Pull request by chrismaree**
- Refactor isUpdated to be a state variable, rather than a private one
- Refactored how pool utilisation is calculated to better separate concerns between hub and RateModelStoreClient clients
- Added method to HubPool client to fetch all L1 tokens and all associated L1 token information include symbol and decimals. This includes a method to get these details for a given deposit.
    - Added tests for ^^
- Added a method to spokePoolClient to fetch if a route is enabled and to fetch all origin tokens from a given spokePool. This will be used in the following PR that adds approval and token balance monitoring logic.
    - Added tests for ^^
- improved logs produced by relayer by leveraging the token symbol and decimal logic added.

### CODE REVIEW

File: src/clients/HubPoolClient.ts [hunk 1]
```diff
@@ -40,37 +41,43 @@ export class HubPoolClient {
...
+    const [current, post] = await Promise.all([
       this.hubPool.callStatic.liquidityUtilizationCurrent(l1Token, blockOffset),
-      this.hubPool.callStatic.liquidityUtilizationPostRelay(l1Token, amount.toString(), blockOffset),
+      this.hubPool.callStatic.liquidityUtilizationPostRelay(l1Token, relaySize, blockOffset),
     ]);
-    return { liquidityUtilizationCurrent, liquidityUtilizationPostRelay };
+    return { current, post };
+  }
+
+  getL1Tokens() {
```
Comment (chrismaree) [inline src/clients/HubPoolClient.ts:57]: new method added in this PR.

**Review by amateima — APPROVED** [review]

File: src/clients/HubPoolClient.ts [hunk 2]
```diff
@@ -40,37 +42,60 @@ export class HubPoolClient {
     return this.l1TokensToDestinationTokens[l1Token][destinationChainId];
   }
 
-  async getPoolUtilization(quoteBlockNumber: number, l1Token: string, amount: BigNumber) {
+  async getCurrentPoolUtilization(l1Token: string) {
+    return await this.hubPool.callStatic.liquidityUtilizationCurrent(l1Token);
+  }
+
+  async getPostRelayPoolUtilization(l1Token: string, quoteBlockNumber: number, relaySize: BigNumber) {
```
Comment (chrismaree) [inline src/clients/HubPoolClient.ts:49]: previously this method was not being used. I've refactored how the `RateModelClient` works to use this method now.

File: src/clients/HubPoolClient.ts [hunk 3]
```diff
@@ -40,37 +42,60 @@ export class HubPoolClient {
...
 
     for (const event of poolRebalanceRouteEvents) {
       const args = spreadEvent(event);
       assign(this.l1TokensToDestinationTokens, [args.l1Token, args.destinationChainId], args.destinationToken);
     }
 
-    this._isUpdated = true;
+    // For each enabled Lp token fetch the token symbol and decimals from the token contract. Note this logic will
+    // only run iff a new token has been enabled. Will only append iff the info is not there already.
+    const tokenInfo = await Promise.all(l1TokensLPEvents.map((event) => this.getTokenInfo(spreadEvent(event).l1Token)));
```
Comment (chrismaree) [inline src/clients/HubPoolClient.ts:87]: fetch token symbol, decimals for each L1 token, iff it was not previously loaded.

File: src/clients/RateModelClient.ts [hunk 4]
```diff
@@ -74,14 +67,9 @@ export class RateModelClient {
...
 
-    this._isUpdated = true;
-
+    this.isUpdated = true;
     this.firstBlockToSearch = searchConfig[1] + 1; // Next iteration should start off from where this one ended.
 
     this.logger.debug({ at: "RateModelClient", message: "Client updated!" });
   }
-
-  isUpdated(): Boolean {
```
Comment (chrismaree) [inline src/clients/RateModelClient.ts:84]: there is no need for this to be a method...can just use the variable as a public getter

File: src/clients/RateModelClient.ts [hunk 5]
```diff
@@ -24,23 +23,19 @@ export class RateModelClient {
...
-
-    const rateModel = this.getRateModelForBlockNumber(l1Token, quoteBlockNumber);
+    const quoteBlock = (await this.blockFinder.getBlockForTimestamp(deposit.quoteTimestamp)).number;
 
-    const blockOffset = { blockTag: quoteBlockNumber };
-    const [utilizationCurrent, utilizationPost] = await Promise.all([
-      this.hubPoolClient.hubPool.callStatic.liquidityUtilizationCurrent(l1Token, blockOffset),
-      this.hubPoolClient.hubPool.callStatic.liquidityUtilizationPostRelay(l1Token, deposit.amount, blockOffset),
-    ]);
+    const { current, post } = await this.hubPoolClient.getPostRelayPoolUtilization(l1Token, quoteBlock, deposit.amount);
```
Comment (chrismaree) [inline src/clients/RateModelClient.ts:28]: now uses the HubPool client method, as this is called on the hub pool.

File: src/clients/SpokePoolClient.ts [hunk 6]
```diff
@@ -35,6 +36,18 @@ export class SpokePoolClient {
...
 
+  getDepositRoutes(): { [originToken: string]: { [DestinationChainId: number]: Boolean } } {
+    return this.depositRoutes;
+  }
+
+  isDepositRouteEnables(originToken: string, destinationChainId: number): boolean {
+    return this.depositRoutes[originToken]?.[destinationChainId] ?? false;
+  }
+
+  getAllOriginTokens(): string[] {
```
Comment (chrismaree) [inline src/clients/SpokePoolClient.ts:47]: can now fetch all "Origin" tokens and if routes are enabled. this is used when working out what chains to preform approvals on.

File: src/relayer/Relayer.ts [hunk 7]
```diff
@@ -50,11 +51,12 @@ export class Relayer {
   }
 
   constructRelayFilledMrkdwn(deposit: Deposit, repaymentChainId: number, amountToFill: BigNumber) {
+    const tokenInfo = this.spokePoolClients[deposit.originChainId].hubPoolClient().getTokenInfoForDeposit(deposit);
```
Comment (chrismaree) [inline src/relayer/Relayer.ts:54]: logs now contain symbol and correctly scaled decimals thanks to the newly added method in the HubPool client.

File: src/clients/SpokePoolClient.ts [hunk 8]
```diff
@@ -35,6 +36,18 @@ export class SpokePoolClient {
     return this.fills;
   }
 
+  getDepositRoutes(): { [originToken: string]: { [DestinationChainId: number]: Boolean } } {
+    return this.depositRoutes;
+  }
+
+  isDepositRouteEnables(originToken: string, destinationChainId: number): boolean {
```
Comment (nicholaspai) [inline src/clients/SpokePoolClient.ts:43]:
```suggestion
  isDepositRouteEnabled(originToken: string, destinationChainId: number): boolean {
```

File: src/utils/ObjectUtils.ts [hunk 9]
```diff
@@ -12,6 +12,8 @@ export function assign(obj: any, keyPath: any[], value: any) {
   if (!obj[keyPath[lastKeyIndex]] || typeof obj[keyPath[lastKeyIndex]] == "string") obj[keyPath[lastKeyIndex]] = value;
   // If the object at the deep path is an array then append array wise.
   else if (Array.isArray(value)) obj[keyPath[lastKeyIndex]] = [...obj[keyPath[lastKeyIndex]], ...value];
+  // If the value is false bool then set to false. This special case is needed as {...false} = {} which causes issues.
+  else if (value === false) obj[keyPath[lastKeyIndex]] = false;
```
Comment (nicholaspai) [inline src/utils/ObjectUtils.ts:16]:
```suggestion
  else if (!value) obj[keyPath[lastKeyIndex]] = false;
```

**Review by nicholaspai — APPROVED** [review]
Nice work! Left some nits

**Pull request merged**